src/components/intent.ts

```diff
@@ -827,23 +850,87 @@ export class Intent {
         return powerLevelEvent;
     }
 
-    private async _ensureRegistered() {
-        if (this.opts.registered) {
+    private async loginForEncryptedClient() {
+        const userId: string = this.client.credentials.userId;
+        const LOGIN_TYPE = "uk.half-shot.unstable.login.appservice";
+        const res = await this.client.login(LOGIN_TYPE, {
+            identifier: {
+                type: "m.id.user",
+                user: userId,
+            }
+        });
+        return {
+            accessToken: res.access_token,
+            deviceId: res.device_id,
+        };
+    }
+
+    public async ensureRegistered() {
+        log.debug("Checking if user is registered");
+        if (this.opts.registered && !this.encryption) {
+            log.debug("ensureRegistered: Registered, and not encrypted");
             return "registered=true";
         }
-        const userId = this.client.credentials.userId;
-        const localpart = new MatrixUser(userId).localpart;
-        try {
-            const res = await this.botClient.register(localpart);
-            this.opts.registered = true;
-            return res;
-        }
-        catch (err) {
-            if (err.errcode === "M_USER_IN_USE") {
+        let registerRes;
+        const userId: string = this.client.credentials.userId;
+        if (!this.opts.registered) {
+            const localpart = new MatrixUser(userId).localpart;
```

Comment:
I wish there was a static function on MatrixUser to get the localpart instead of creating an object just to throw it away. Also, the priority of new is basically never talked about. I first read this as `new (MatrixUser(userId).localpart)`
```suggestion
            const localpart = (new MatrixUser(userId)).localpart;
```

Reply:
Have fixed the priority. Agreed I wish this was static.